pyrobosim/pyrobosim/navigation/execution.py

```diff
@@ -205,6 +251,65 @@ def validate_remaining_path(self) -> None:
 
             time.sleep(max(0, self.validation_dt - (time.time() - start_time)))
 
+    def detect_closed_hallway(self) -> None:
+        """
+        Get lidar measurement and determine if it's scanning a hallway.
+        If yes, it would update the robot recorded_closed_hallways knowledge.
+        It either remove (if it detects hallway is open) or add (if it detects hallway is close) the hallway
+        into the robot's knowledge.
+        """
+        if (self.robot is None) or (self.robot.world is None):
+            return
+
+        lidar_sensor = self.robot.sensors.get(self.lidar_sensor_name)  # type: ignore[arg-type, union-attr]
+        measured_angles = lidar_sensor.angles  # type: ignore[union-attr]
+
+        while self.following_path and (not self.abort_execution):
+            start_time = time.time()
+            cur_pose = self.robot.get_pose()
+
+            measured_lengths = lidar_sensor.get_measurement()  # type: ignore[union-attr]
+            analyse_pose = []
+            for angle, length in zip(measured_angles, measured_lengths):
+                if length < lidar_sensor.max_range_m:  # type: ignore[union-attr]
+                    # There are objects in lidar line of sight
+                    adjusted_angle = angle + cur_pose.get_yaw()
+                    x = cur_pose.x + length * np.cos(adjusted_angle)
+                    y = cur_pose.y + length * np.sin(adjusted_angle)
+                    analyse_pose.append((x, y))
```

Comment:
I think you can probably speed this up dramatically with list comprehensions. Something like this:

```python
yaw = cur_pose.get_yaw()
analyse_poses = [
    (
        cur_pose.x + length * np.cos(angle + yaw),
        cur_pose.y + length * np.sin(angle + yaw),
    )
    for angle, length in zip(measured_angles, measured_lengths)
]
```